internal/document.go

```diff
@@ -54,3 +55,189 @@ func (doc Document) Debug() (bool, error) {
 	}
 	return true, nil
 }
+
+func (doc Document) Verify(
+	certProvider CertProvider,
+	verificationTime VerificationTimeFunc,
+) (*x509.Certificate, []*x509.Certificate, error) {
+	err := doc.CheckMandatoryFields()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = doc.CheckOptionalFields()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cert, certificates, err := doc.CheckCertificates(certProvider, verificationTime)
+	if err != nil {
+		return nil, nil, err
+	}
+	return cert, certificates, nil
+}
+
+func missingFieldError(field string) error {
+	return fmt.Errorf("missing %s", field)
+}
+
+func (doc Document) CheckMandatoryFields() error {
+	if doc.ModuleID == "" {
+		return missingFieldError("module id")
+	}
+	if doc.Digest == "" {
+		return missingFieldError("digest")
+	}
+	if doc.Timestamp < 1 {
+		return missingFieldError("timestamp")
+	}
+	if doc.PCRs == nil {
+		return missingFieldError("pcrs")
+	}
+	if doc.Certificate == nil {
+		return missingFieldError("certificate")
+	}
+	if doc.Digest != "SHA384" {
+		return fmt.Errorf("expected 'SHA384' digest but got '%s' digest", doc.Digest)
+	}
+	if len(doc.PCRs) < 1 || len(doc.PCRs) > 32 {
+		return fmt.Errorf("expected 1 to 32 pcrs but got %v pcrs", len(doc.PCRs))
+	}
+
+	for key, value := range doc.PCRs {
+		if key > 31 {
+			return fmt.Errorf("pcr key '%v' is out of range [0, 31]", key)
+		}
+
+		if value == nil {
+			return fmt.Errorf("pcr value for key '%v' is nil", key)
+		}
+
+		if !(32 == len(value) ||
+			48 == len(value) ||
+			64 == len(value) ||
+			96 == len(value)) {
+			return fmt.Errorf(
+				"expected pcr len of 32, 48, 64, or 96 but got '%v'",
+				len(value),
+			)
+		}
+	}
+	return nil
+}
+
+func (doc Document) CheckOptionalFields() error {
+	if len(doc.PublicKey) > MaxPublicKeyLen {
+		return fmt.Errorf(
+			"max public key len is '%v' but got '%v'",
+			MaxPublicKeyLen,
+			len(doc.PublicKey),
+		)
+	}
+
+	if len(doc.UserData) > MaxUserDataLen {
+		return fmt.Errorf(
+			"max user data len is '%v' but got '%v'",
+			MaxUserDataLen,
+			len(doc.UserData),
+		)
+	}
+
+	if len(doc.Nonce) > MaxNonceLen {
+		return fmt.Errorf(
+			"max nonce len is '%v' but got '%v'",
+			MaxNonceLen,
+			len(doc.Nonce),
+		)
+	}
+	return nil
+}
+
+func (doc Document) CheckCertificates(
+	certProvider CertProvider,
+	verificationTime VerificationTimeFunc,
+) (*x509.Certificate, []*x509.Certificate, error) {
```

Comment:
You'll notice below that the first certificate in the chain is always the `doc.Certificate` which gets used later in `internal/nitrite.Verify()` to verify the signature.

Initially, I tried returning the array of certificates and had `nitrite.Verify()` use the first element for signature verification, but this would lead to a panic during testing. I noticed that the panic only ever happened when verifying the self-signed cert (nitro attestations worked fine). Mike said the self-signed cert code is somewhat "janky". So, instead of trying to fix it, I decided to return the `doc.Certificate` separately which seems to solve the issue.

Reply:
I'm not seeing that error. Maybe double check on your end? Check out the two suggestions I made to `nitrite.go` - I think you can just return `certificates`

Reply:
Hmmmm weird. I tried it out as you said and it seems to work fine... Would like to know why it wasn't working previously, but at least it works now. Updated